File: app/rotas/validacao.py
"""/validar — a página pública que confirma um certificado pela chave.

Fatia 5 do desenho de Certificados, e a única rota deste sistema que **não pede
sessão**. A exceção é declarada aqui, testada em
`testes/integracao/test_validacao_publica.py` e conferida pelo invariante
`test_so_estas_rotas_abrem_sem_sessao`: negado por padrão continua valendo, e o
que muda é que a lista das exceções passou a existir por escrito.

**Por que ela existe agora.** `emissao_certificado.montar_contexto_para_emitir`
grava `url_validacao = "{url_base}/validar/{chave}"` em todo certificado, o
docxtpl a imprime no papel e no QR, e a ficha a exibe. A rota não existia: cada
certificado que saiu do setor circula com um endereço que responde 404, e papel
não se corrige depois de entregue.

**Por que sem login.** A página serve a quem NÃO tem conta — o órgão, a banca de
concurso, o fiscal de contrato, a empresa que recebeu o certificado e quer saber
se ele é autêntico. Quem tem conta aqui já abre `/certificados` ou
`/certificados/meus`, e é justamente quem não precisa dela. Exigir login não
tornaria a página mais segura: tornaria-a inútil, porque o único público que ela
tem ficaria de fora.

**O que ela custa hoje: nada de novo.** A rota é alcançável exatamente de onde
todas as outras já são — a rede do setor —, e publicar o sistema na internet
continua sendo decisão institucional a tomar com o TI, registrada como tal no
`docs/ROPA.md` §6. Esta rota não a toma nem a antecipa; ela garante que, no dia
em que for tomada, o endereço impresso responda.

**O que ela nunca mostra:** o nome de quem se formou. A regra e o argumento moram
em `servicos/validacao_certificado.py`, junto do limite por IP e da conferência
de nome que entra no lugar da exibição.

A tela sai pela mesma casca das telas anteriores à sessão (`sem_casca`), e não
por um segundo layout: `base.html` já tem o interruptor, com o argumento escrito
lá — um `base_publica.html` duplicaria `<head>`, rodapé e o salto para o conteúdo
para tirar uma `<aside>` que, sem `usuario`, já não é desenhada.
"""

from __future__ import annotations

from datetime import date

from fastapi import APIRouter, Form, Request
from fastapi.responses import RedirectResponse

from app.dependencias import SessaoDep
from app.servicos import validacao_certificado as servico
from app.servicos.certificado import normalizar_chave
from app.web import pagina

rotas = APIRouter(tags=["validacao"])

VALIDAR = "/validar"

# A chave vai na URL porque é assim que o QR do papel a carrega. O `maxlength`
# do formulário e o corte aqui são o mesmo número: chave é de tamanho fixo, e um
# caminho de 4 KB só serviria para encher o registro de acesso.
TAMANHO_MAXIMO_CHAVE = 40


def _sem_rastro(resposta):
    """`noindex` e `no-store` na resposta pública.

    `no-store` porque a página responde sobre um documento de uma pessoa e não
    pode ficar no cache do navegador de um balcão compartilhado; `noindex`
    porque uma página de validação indexada convida exatamente a varredura que o
    limite por IP existe para conter. Vão aqui e não no middleware de segurança:
    lá o `setdefault` vale para toda resposta do sistema, e `no-store` em toda
    tela autenticada seria uma decisão de desempenho tomada de lado.
    """
    resposta.headers["x-robots-tag"] = "noindex, nofollow"
    resposta.headers["cache-control"] = "no-store"
    return resposta


def _tela(
    request: Request,
    *,
    resposta: servico.Resposta | None = None,
    digitada: str = "",
    confere: bool | None = None,
    nome_digitado: str = "",
    excedeu: bool = False,
    status: int = 200,
):
    saida = pagina(
        request,
        "paginas/validacao.html",
        usuario=None,
        resposta=resposta,
        digitada=digitada,
        confere=confere,
        nome_digitado=nome_digitado,
        excedeu=excedeu,
        hoje=date.today(),
        limite_curto=servico.LIMITE_CURTO,
    )
    saida.status_code = status
    return _sem_rastro(saida)


@rotas.get(VALIDAR)
def formulario(request: Request):
    """A tela sem chave: só o campo e a explicação do que ela responde."""
    return _tela(request)


@rotas.post(VALIDAR)
def enviar(request: Request, chave: str = Form("")):
    """O campo digitado vira URL.

    Redireciona em vez de responder no POST para que o endereço da resposta seja
    o MESMO que está impresso no papel — quem conferiu por digitação e quem leu o
    QR terminam na mesma URL, e ela pode ser copiada para um processo.
    """
    limpa = normalizar_chave(chave)[:TAMANHO_MAXIMO_CHAVE]
    if not limpa:
        return _sem_rastro(RedirectResponse(VALIDAR, status_code=303))
    return _sem_rastro(RedirectResponse(f"{VALIDAR}/{limpa}", status_code=303))


@rotas.get(VALIDAR + "/{chave}")
def validar(request: Request, s: SessaoDep, chave: str):
    """A resposta: autêntico ou não, e nunca de quem.

    O limite por IP é cobrado ANTES da consulta, e conta a tentativa mesmo quando
    a chave é válida: contar só o erro convidaria a varrer com chaves conhecidas.
    """
    ip = request.client.host if request.client else None
    if not servico.dentro_do_limite(ip):
        # 429, e a tela diz o que fazer. Um 403 aqui misturaria "você excedeu" com
        # "você não pode", que são coisas diferentes para quem só digitou errado.
        return _tela(request, digitada=chave, excedeu=True, status=429)
    resposta = servico.consultar(s, chave[:TAMANHO_MAXIMO_CHAVE])
    return _tela(request, resposta=resposta, digitada=normalizar_chave(chave))


@rotas.post(VALIDAR + "/{chave}/conferir")
def conferir(request: Request, s: SessaoDep, chave: str, nome: str = Form("")):
    """Confere / não confere — um bit, no lugar do nome.

    Passa pelo mesmo limite por IP da consulta, e é isso que torna inviável
    reconstruir um nome por tentativa: sem o limite, um bit por requisição ainda
    é um oráculo.
    """
    ip = request.client.host if request.client else None
    if not servico.dentro_do_limite(ip):
        return _tela(request, digitada=chave, excedeu=True, status=429)
    limpa = normalizar_chave(chave)[:TAMANHO_MAXIMO_CHAVE]
    resposta = servico.consultar(s, limpa)
    confere = servico.confere_o_nome(s, limpa, nome) if resposta.encontrado else None
    return _tela(
        request,
        resposta=resposta,
        digitada=limpa,
        confere=confere,
        # o que foi digitado volta para a tela: sem isso, "não confere" some junto
        # com o texto que o produziu e a pessoa não sabe se errou a digitação
        nome_digitado=(nome or "").strip(),
    )

File: nuvem/src/rotas/validacao.ts
/**
 * /validar — a página pública que confirma um certificado pela chave.
 * Porte de `app/rotas/validacao.py`.
 *
 * A única rota do módulo que **não pede sessão**, como no Python: a página
 * serve a quem NÃO tem conta — o órgão, a banca, a empresa que recebeu o
 * certificado e quer saber se ele é autêntico. O endereço vem impresso no
 * papel e no QR (`emissao_certificado.url_base_validacao`), e papel não se
 * corrige depois de entregue.
 *
 * **O que ela nunca mostra:** o nome de quem se formou. A regra mora em
 * `servicos/validacao_certificado.ts`, junto do limite por IP e da conferência
 * de nome que entra no lugar da exibição.
 */
import { Hono } from "hono";
import type { Ambiente, Ctx } from "../nucleo/contexto.js";
import { formulario } from "../dependencias.js";
import { normalizar_chave } from "../servicos/certificado.js";
import * as datas_br from "../servicos/datas_br.js";
import * as servico from "../servicos/validacao_certificado.js";
import { pagina, redirecionar } from "../web.js";
import { ipDoCliente } from "./autenticacao.js";

export const rotas = new Hono<Ambiente>();

export const VALIDAR = "/validar";

// A chave vai na URL porque é assim que o QR a carrega; um caminho de 4 KB só
// serviria para encher o registro de acesso.
export const TAMANHO_MAXIMO_CHAVE = 40;

/**
 * `noindex` e `no-store` na resposta pública: a página responde sobre um
 * documento de uma pessoa e não pode ficar no cache de um balcão
 * compartilhado; indexada, convidaria a varredura.
 */
function _sem_rastro(resposta: Response): Response {
  resposta.headers.set("x-robots-tag", "noindex, nofollow");
  resposta.headers.set("cache-control", "no-store");
  return resposta;
}

async function _tela(
  c: Ctx,
  d: {
    resposta?: servico.Resposta | null;
    digitada?: string;
    confere?: boolean | null;
    nome_digitado?: string;
    excedeu?: boolean;
    status?: number;
  } = {},
): Promise<Response> {
  const saida = await pagina(
    c,
    "paginas/validacao.html",
    null,
    {
      resposta: d.resposta ?? null,
      digitada: d.digitada ?? "",
      confere: d.confere ?? null,
      nome_digitado: d.nome_digitado ?? "",
      excedeu: d.excedeu ?? false,
      hoje: datas_br.hoje(),
      limite_curto: servico.LIMITE_CURTO,
    },
    d.status ?? 200,
  );
  return _sem_rastro(saida);
}

/** A tela sem chave: só o campo e a explicação do que ela responde. */
rotas.get(VALIDAR, (c) => _tela(c));

/**
 * O campo digitado vira URL: o endereço da resposta é o MESMO que está
 * impresso no papel.
 */
rotas.post(VALIDAR, async (c) => {
  const f = await formulario(c);
  const limpa = normalizar_chave(f.texto("chave")).slice(0, TAMANHO_MAXIMO_CHAVE);
  if (!limpa) return _sem_rastro(redirecionar(c, VALIDAR));
  return _sem_rastro(redirecionar(c, `${VALIDAR}/${encodeURIComponent(limpa)}`));
});

/**
 * A resposta: autêntico ou não, e nunca de quem. O limite por IP é cobrado
 * ANTES da consulta, e conta a tentativa mesmo quando a chave é válida.
 */
rotas.get(`${VALIDAR}/:chave`, async (c) => {
  const chave = c.req.param("chave");
  if (!servico.dentro_do_limite(ipDoCliente(c))) {
    // 429, e a tela diz o que fazer — "você excedeu" não é "você não pode"
    return _tela(c, { digitada: chave, excedeu: true, status: 429 });
  }
  const resposta = await servico.consultar(c.get("tx"), chave.slice(0, TAMANHO_MAXIMO_CHAVE));
  return _tela(c, { resposta, digitada: normalizar_chave(chave) });
});

/**
 * Confere / não confere — um bit, no lugar do nome. Passa pelo mesmo limite
 * por IP: sem ele, um bit por requisição ainda é um oráculo.
 */
rotas.post(`${VALIDAR}/:chave/conferir`, async (c) => {
  const chave = c.req.param("chave");
  const f = await formulario(c);
  const nome = f.texto("nome");
  if (!servico.dentro_do_limite(ipDoCliente(c))) {
    return _tela(c, { digitada: chave, excedeu: true, status: 429 });
  }
  const tx = c.get("tx");
  const limpa = normalizar_chave(chave).slice(0, TAMANHO_MAXIMO_CHAVE);
  const resposta = await servico.consultar(tx, limpa);
  const confere = resposta.encontrado ? await servico.confere_o_nome(tx, limpa, nome) : null;
  return _tela(c, {
    resposta,
    digitada: limpa,
    confere,
    // o que foi digitado volta: sem isso "não confere" some junto com o texto
    nome_digitado: nome.trim(),
  });
});
